jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java

```diff
@@ -57,6 +58,64 @@
     private boolean _expect100Continue;
     private boolean _delayedUntilContent;
     private boolean _useOutputDirectByteBuffers;
+    private final RequestContent _requestContent = new RequestContent();
+
+    private class RequestContent
+    {
+        private HttpInput.Content _content;
+        private boolean _endStream;
+        private boolean _producing;
+        private final AutoLock _lock = new AutoLock();
+
+        void setContent(boolean endStream, HttpInput.Content content)
+        {
+            try (AutoLock ignored = _lock.lock())
+            {
+                if (_content != null)
+                    throw new AssertionError("content cannot be queued; stream=" + getStream());
+                _endStream = endStream;
+                _content = content;
+                _producing = false;
+            }
+        }
+
+        private HttpInput.Content takeContent(boolean[] endStreamResult)
+        {
+            try (AutoLock ignored = _lock.lock())
+            {
+                if (_content == null)
+                    return null;
+                HttpInput.Content contentCopy = _content;
+                endStreamResult[0] = _endStream;
+                _content = null;
+                _endStream = false;
+                return contentCopy;
+            }
+        }
+
+        HttpInput.Content takeContentOrDemand(boolean[] endStreamResult)
```

Comment:
I'm not a big fan of the passed array!   I'd much rather you return a `HttpInput.LastContent` wrapper of the `HttpInput.Content`

Reply:
This method actually used to return `Map.Entry<Boolean, HttpInput.Content>` but @sbordet preferred to mimic the `j.u.c.a.AtomicMarkableReference.get(boolean[])` contract.

Reply:
hhhhmmmm maybe....  OK leave as is for now....   but a LastContent wrapper could be better and would be less allocations.  This allocates an array for every call, the Map.Entry allocated for every return.  The LastContent wrapper would only allocated for the last return!